packages/fast-components-react-base/src/tabs/tabs.tsx

```diff
@@ -0,0 +1,313 @@
+                    aria-selected={this.state.activeId !== id ? false : true}
+                    onClick={this.handleClick}
+                    onKeyDown={this.handleKeyDown}
+                    tabIndex={this.state.activeId !== id ? -1 : 0}
+                >
+                    {this.getChildBySlot(tabItem.props.children, TabSlot.tab)[0]}
+                </button>
+            );
+        });
+
+        return tabElements;
+    }
+
+    /**
+     * Renders the tab panels
+     */
+    private renderTabPanels(): JSX.Element[] {
+        const tabPanels: JSX.Element[] = [];
+
+        this.getChildBySlot(this.props.children, TabSlot.tabItem).forEach((tabItem: JSX.Element, index: number): void => {
+            const id: string = get(tabItem, "props.id") ? tabItem.props.id : `tab${index}`;
+
+            tabPanels.push(
+                <div
+                    key={id}
+                    id={id}
+                    className={this.props.managedClasses.tab_panel}
+                    role="tabpanel"
+                    aria-labelledby={id}
+                    aria-hidden={this.state.activeId !== id ? true : false}
+                >
+                    {this.getChildBySlot(tabItem.props.children, TabSlot.tabPanel)[0]}
+                </div>
+            );
+        });
+
+        return tabPanels;
+    }
+
+    /**
+     * Generate class names for tab elements
+     */
+    private generateTabClassNames(id: string): string {
+        return this.state.activeId !== id
+            ? this.props.managedClasses.tab
+            : `${this.props.managedClasses.tab} ${this.props.managedClasses.tab__active}`;
+    }
+
+    /**
+     * Handles the click event on the tab element
+     */
+    private handleClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
+        if (!this.props.activeId) {
+            this.setState({
+                activeId: e.currentTarget.getAttribute("aria-controls")
+            });
+        } else if (typeof this.props.onUpdateTab === "function") {
+            this.props.onUpdateTab(e.currentTarget.getAttribute("aria-controls"));
+        }
+    }
+
+    /**
+     * Handles the keydown event on the tab element
+     */
+    private handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>): void => {
+        switch (e.keyCode) {
+            case KeyCodes.arrowLeft:
+            case KeyCodes.arrowUp:
+                this.activatePrevious();
+                break;
+            case KeyCodes.arrowRight:
+            case KeyCodes.arrowDown:
+                this.activateNext();
+                break;
+            case KeyCodes.home:
+                this.activateFirst();
+                break;
+            case KeyCodes.end:
+                this.activateLast();
+                break;
+        }
+    }
+
+    /**
+     * Activates the previous tab item
+     */
+    private activatePrevious(): void {
+        const items: JSX.Element[] = this.getChildBySlot(this.props.children, TabSlot.tabItem);
```

Comment:
There is a decent amount of duplication between `activePrevious`, `activateNext`, `activateFirst`, and `activateLast` - I think you could refactor to a single function that shifts the `activeId` by a particular value instead: `shiftActiveTab(-1)`, `shiftActiveTab(1)`, `shiftActiveTab(this.state.items.length - this.getChildBySlot(this.props.children, TabSlot.tabItem)`, etc